ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
 
-  rb_ary_push(server_list, rb_server);
-  return MEMCACHED_SUCCESS;
+	rb_ary_push(rb_server_list, rb_server);
+	return MEMCACHED_SUCCESS;
 }
 
 static VALUE
 rb_connection_servers(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_server_fn callbacks[1] = { &iterate_server_function };
+	memcached_st *mc;
 
-  VALUE server_list = rb_ary_new();
+	VALUE rb_result;
 
-  memcached_server_fn callbacks[1];
-  callbacks[0] = iterate_server_function;
+	UnwrapMemcached(self, mc);
+	rb_result = rb_ary_new();
 
-  memcached_server_cursor(ctx->memc, callbacks, (void *)server_list, 1);
-  return server_list;
+	memcached_server_cursor(mc, callbacks, (void *)rb_result, 1);
+	return rb_result;
 }
 
 static VALUE
 rb_connection_flush(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_st *mc;
+	memcached_return_t rc;
 
-  memcached_return_t rc = memcached_flush(ctx->memc, 0);
-  handle_memcached_return(rc);
+	UnwrapMemcached(self, mc);
+	rc = memcached_flush(mc, 0);
+	rb_memcached_error_check(rc);
 
-  return (rc == MEMCACHED_SUCCESS);
+	return (rc == MEMCACHED_SUCCESS) ? Qtrue : Qfalse;
 }
 
 static VALUE
-rb_connection_set(VALUE self, VALUE key, VALUE value, VALUE ttl, VALUE flags)
+rb_connection_set(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(key);
-  char *mvalue = StringValuePtr(value);
-
-  memcached_return_t rc = memcached_set(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(ttl), NUM2INT(flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_set(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+			
+	rb_memcached_return(rc);
 }
 
 static VALUE
-rb_connection_get(VALUE self, VALUE key)
+rb_connection_get(VALUE self, VALUE rb_key)
 {
-  Check_Type(key, T_STRING);
+	VALUE rb_val;
+
+	memcached_st *mc;
+	memcached_return_t rc;
+	size_t ret_len;
+	uint32_t ret_flags;
+	char *ret;
 
-  memcached_ctx *ctx = get_ctx(self);
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
 
-  char *mkey = StringValuePtr(key);
+	ret = memcached_get(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		&ret_len, &ret_flags, &rc);
 
-  size_t return_value_length;
-  memcached_return_t rc;
-  uint32_t flags; // TODO return flags so client can know if it should deserialize or decompress
-  char *response  = memcached_get(ctx->memc, mkey, strlen(mkey), &return_value_length, &flags, &rc);
+	if (ret == NULL) {
+		rb_memcached_error_check(rc);
+		return Qnil;
+	}
 
-  handle_memcached_return(rc);
-  if (rc == MEMCACHED_SUCCESS)
-    return rb_str_new2(response);
-  else
-    return Qnil;
+	rb_val = rb_str_new(ret, ret_len);
```

Comment:
Careful with the return value of `memcached_get`! If it's non-NULL, it's a library-allocated string which we **must** free. Calling `rb_str_new` keeps a copy of the string, so we were always leaking on all `get` operations! Scary!

Also note we never `rb_str_new2`. We always pass an explicit length for the strings we create, again, because of NULL bytes. The `return_value_length` we get from memcached is there for that exact reason. `rb_str_new2` was calling `strlen` on the string you passed, so all gets for binary data would have been the wrong length. Also very scary!